src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,305 @@
+/-
+Copyright (c) 2021 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Eric Wieser
+-/
+
+import ring_theory.ideal.basic
+import ring_theory.ideal.operations
+import linear_algebra.finsupp
+import ring_theory.graded_algebra.basic
+
+/-!
+
+# Homogeneous ideal of a graded algebra
+
+This file defines homogeneous ideals of `graded_algebra A` where `A : ι → submodule R R`and
+operations on them:
+* `mul`, `inf`, `Inf` of homogeneous ideals are homogeneous;
+* `⊤`, `⊥`, i.e. the trivial ring and `R` are homogeneous;
+* `radical` of a homogeneous ideal is homogeneous.
+-/
+
+noncomputable theory
+
+section is_homogeneous_ideal_defs
+
+open set_like direct_sum set
+open_locale big_operators
+
+variables {ι R : Type*} [comm_ring R] [decidable_eq ι] [add_comm_monoid ι]
+variables (A : ι → submodule R R) [graded_algebra A]
+variable (I : ideal R)
+
+/--An `I : ideal R` is called homogeneous if for every `r ∈ I`, every homogeneous component of `r`
+  is in `I`.-/
+def is_homogeneous_ideal : Prop :=
+∀ ⦃i : ι⦄ ⦃r : R⦄, r ∈ I → (graded_algebra.decompose A r i : R) ∈ I
+
+lemma is_homogeneous_ideal_iff_forall_subset :
+  is_homogeneous_ideal A I ↔ ∀ i, (I : set R) ⊆ graded_algebra.proj A i ⁻¹' I :=
+iff.rfl
+
+lemma is_homogeneous_ideal_iff_subset_Inter :
+  is_homogeneous_ideal A I ↔ (I : set R) ⊆ ⋂ i, graded_algebra.proj A i ⁻¹' ↑I :=
+subset_Inter_iff.symm
+
+lemma is_homogeneous_ideal.exists_iff_eq_span :
+  (∃ (S : set (homogeneous_submonoid A)), I = ideal.span (coe '' S)) ↔
+  I = ideal.span {x | x ∈ I ∧ is_homogeneous A x} :=
+-- get rid of the messy subtypes and set coercions
+suffices (∃ s : set R, s ⊆ set_of (is_homogeneous A) ∧ I = ideal.span s) ↔
+  I = ideal.span (I ∩ set_of (is_homogeneous A)),
+from (subtype.exists_set_subtype _).trans this,
+begin
+  split,
+  { rintros ⟨s, hs, rfl⟩,
+    apply le_antisymm,
+    { exact ideal.span_mono (subset_inter (ideal.subset_span) hs) },
+    { exact ideal.span_le.2 (inter_subset_left _ _) } },
+  { intros hI,
+    exact ⟨(I : set R) ∩ set_of (is_homogeneous A), inter_subset_right _ _, hI⟩, }
+end
```

Comment:
I golfed this substantially - you can now see that actually this isn't really a lemma about `is_homogeneous` at all, once you eliminate the subtypes and coercions, `set_of (is_homogeneous A)` could be replaced with any set.

Reply:
I guess this is still TODO

Reply:
Yes, I think we might be proving things in the wrong order and spending lots of work juggling awkward definitions.